### PR TITLE
*: optimize error generation, debug log and AllocHandle in the hot write path

### PR DESCRIPTION
**Pull request by ekexium**
### What problem does this PR solve?

Issue Number: ref #54444

Problem Summary:

#### debug logging
When running `BenchmarkPipelinedInsertIgnore`, here are 3 versions of the same logging. The vanilla code takes noticeable time.
```
     150ms      1.95s    936:	logutil.BgLogger().Debug("alloc N signed ID",
         .       20ms    937:		zap.Uint64("from ID", uint64(alloc.base)),
         .      180ms    938:		zap.Uint64("to ID", uint64(alloc.base+n1)),
         .       10ms    939:		zap.Int64("table ID", alloc.tbID),
         .          .    940:		zap.Int64("database ID", alloc.dbID))
         .      300ms    941:	if ce := logutil.BgLogger().Check(zap.DebugLevel, "alloc N signed ID"); ce != nil {
         .          .    942:		ce.Write(
         .          .    943:			zap.Uint64("from ID", uint64(alloc.base)),
         .          .    944:			zap.Uint64("to ID", uint64(alloc.base+n1)),
         .          .    945:			zap.Int64("table ID", alloc.tbID),
         .          .    946:			zap.Int64("database ID", alloc.dbID),
         .          .    947:		)
         .          .    948:	}
      30ms      160ms    949:	if logutil.BgLogger().Core().Enabled(zap.DebugLevel) {
         .          .    950:		logutil.BgLogger().Debug("alloc N signed ID",
         .          .    951:			zap.Uint64("from ID", uint64(alloc.base)),
         .          .    952:			zap.Uint64("to ID", uint64(alloc.base+n1)),
         .          .    953:			zap.Int64("table ID", alloc.tbID),
         .          .    954:			zap.Int64("database ID", alloc.dbID))
         .          .    955:	}
```

#### returning an interface
Previous (line 1750):
```
     180ms      1.26s (flat, cum)  0.96% of Total
      40ms       40ms   1737:func AllocHandle(ctx context.Context, mctx table.MutateContext, t table.Table) (kv.Handle, error) {
         .          .   1738:	if mctx != nil {
      60ms       70ms   1739:		if stmtCtx := mctx.GetSessionVars().StmtCtx; stmtCtx != nil {
         .          .   1740:			// First try to alloc if the statement has reserved auto ID.
      50ms       50ms   1741:			if stmtCtx.BaseRowID < stmtCtx.MaxRowID {
         .          .   1742:				stmtCtx.BaseRowID++
         .          .   1743:				return kv.IntHandle(stmtCtx.BaseRowID), nil
         .          .   1744:			}
         .          .   1745:		}
         .          .   1746:	}
         .          .   1747:
      20ms      700ms   1748:	_, rowID, err := AllocHandleIDs(ctx, mctx, t, 1)
         .          .   1749:	ret := kv.IntHandle(rowID)
      10ms      400ms   1750:	return ret, err
         .          .   1751:}
```

This PR (line 1751):
```
     110ms      900ms (flat, cum)  0.69% of Total
      30ms       30ms   1737:func AllocHandle(ctx context.Context, mctx table.MutateContext, t table.Table) (kv.IntHandle,
         .          .   1738:	error) {
      20ms       20ms   1739:	if mctx != nil {
      20ms       20ms   1740:		if stmtCtx := mctx.GetSessionVars().StmtCtx; stmtCtx != nil {
         .          .   1741:			// First try to alloc if the statement has reserved auto ID.
      10ms       10ms   1742:			if stmtCtx.BaseRowID < stmtCtx.MaxRowID {
         .          .   1743:				stmtCtx.BaseRowID++
      10ms       10ms   1744:				return kv.IntHandle(stmtCtx.BaseRowID), nil
         .          .   1745:			}
         .          .   1746:		}
         .          .   1747:	}
         .          .   1748:
      10ms      800ms   1749:	_, rowID, err := AllocHandleIDs(ctx, mctx, t, 1)
         .          .   1750:	ret := kv.IntHandle(rowID)
      10ms       10ms   1751:	return ret, err
         .          .   1752:}
```

#### error generation
In the `BenchmarkPipelinedInsertOnDuplicate` scenario, the error generation takes more than 5% of CPU time.
Previous:
```
      20ms      390ms    214:			uniqueKeys = append(uniqueKeys, &keyValueWithDupInfo{
      10ms       10ms    215:				newKey: key,
     140ms      3.99s    216:				dupErr: kv.ErrKeyExists.FastGenByArgs(colValStr, fmt.Sprintf("%s.%s", v.TableMeta().Name.String(), v.Meta().Name.String())),
         .          .    217:			})
```

This PR:
```
      20ms      430ms    213:			uniqueKeys = append(uniqueKeys, &keyValueWithDupInfo{
      30ms       30ms    214:				newKey: key,
      30ms      1.73s    215:				dupErr: kv.ErrKeyExists.FastGenByArgs(colValStr,
      90ms      1.10s    216:					v.TableMeta().Name.String()+"."+v.Meta().Name.String()),
         .          .    217:			})
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/session/bench_test.go [hunk 1]
```diff
@@ -1944,3 +1945,65 @@ func TestBenchDaily(t *testing.T) {
...
+	se.GetSessionVars().StmtCtx.InInsertStmt = true
+	for i := 0; i < b.N; i++ {
+		se.Execute(context.Background(), "insert ignore into tmp select * from src")
+		warningCount := se.GetSessionVars().StmtCtx.WarningCount()
+		require.Zero(b, warningCount)
+	}
+	b.StopTimer()
+}
+
+func BenchmarkPipelinedInsertOnDuplicate(b *testing.B) {
```
Comment (hawkingrei) [inline pkg/session/bench_test.go:1977]: Can you add this benchmark to dailybench?

Comment (ekexium) [reply pkg/session/bench_test.go:1977]: I didn't do this because I don't know how it's maintained. @tiancaiamao how do you like it?

**Review by zyguan — APPROVED** [review]

**Review by cfzjywxk — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**